From 11ecce4405ea27e4c7bdc1fda6a3f6f4dd66e6cc Mon Sep 17 00:00:00 2001
From: Hou Qi <qi.hou@nxp.com>
Date: Fri, 6 Dec 2024 13:21:35 +0800
Subject: [PATCH 72/75] LF-14072 v4l2: fix downscale has abnormal lines at
 bottom

Upstream-Status: Pending

When downscaling for videos that have crop info, wave6 hardware won't
crop first before downscale, it uses whole buffer including extra
padding lines. Then display has abnormal lines at bottom.

This patch adjusts downscale size according to padding info. It
adds extra width and extra height when downscaling for videos that
needs padding. Then adds these extra size to cropmeta to meet the
actual downscale size requirement.

This patch also enables downscale playback of videos that have left
and top crop info shows correctly.
Signed-off-by: Xiaolei Wang <xiaolei.wang@windriver.com>
---
 sys/v4l2/gstv4l2object.c   |  83 ++++++++++++++++++++++---
 sys/v4l2/gstv4l2object.h   |   5 +-
 sys/v4l2/gstv4l2videodec.c | 122 ++++++++++++++++---------------------
 3 files changed, 129 insertions(+), 81 deletions(-)

diff --git a/sys/v4l2/gstv4l2object.c b/sys/v4l2/gstv4l2object.c
index 27d2e77..746e4e0 100644
--- a/sys/v4l2/gstv4l2object.c
+++ b/sys/v4l2/gstv4l2object.c
@@ -581,8 +581,9 @@ gst_v4l2_object_new (GstElement * element,
 
   v4l2object->poll = gst_poll_new (TRUE);
   v4l2object->can_poll_device = TRUE;
-  v4l2object->crop_width = 0;
-  v4l2object->crop_height = 0;
+  v4l2object->downstream_width = 0;
+  v4l2object->downstream_height = 0;
+  v4l2object->downscale = FALSE;
   g_atomic_int_set (&v4l2object->seek, FALSE);
 
   /* We now disable libv4l2 by default, but have an env to enable it. */
@@ -4540,6 +4541,32 @@ gst_v4l2_object_try_format (GstV4l2Object * v4l2object, GstCaps * caps,
   return gst_v4l2_object_set_format_full (v4l2object, caps, TRUE, error);
 }
 
+static void
+gst_v4l2_object_adjust_downscale_size (GstV4l2Object * v4l2object,
+    struct v4l2_format * fmt, struct v4l2_rect * r, guint * wanted_width,
+    guint * wanted_height)
+{
+  guint src_width, src_height;
+  guint padded_width, padded_height;
+  guint downscale_width, downscale_height;
+
+  src_width = v4l2object->info.width;
+  src_height = v4l2object->info.height;
+  padded_width = fmt->fmt.pix.width;
+  padded_height = fmt->fmt.pix.height;
+  downscale_width = v4l2object->downstream_width;
+  downscale_height = v4l2object->downstream_height;
+
+  *wanted_width = padded_width * downscale_width / src_width;
+  *wanted_height = padded_height * downscale_height / src_height;
+
+  r->left = r->left * (*wanted_width) / padded_width;
+  r->top = r->top * (*wanted_height) / padded_height;
+
+  GST_INFO_OBJECT (v4l2object->dbg_obj, "downscale size adjusted to %dx%d",
+      *wanted_width, *wanted_height);
+}
+
 /**
  * gst_v4l2_object_acquire_format:
  * @v4l2object: the object
@@ -4562,6 +4589,7 @@ gst_v4l2_object_acquire_format (GstV4l2Object * v4l2object, GstVideoInfo * info)
   guint width, height;
   GstVideoAlignment align;
   GstVideoInterlaceMode interlace_mode;
+  gint crop_width = 0, crop_height = 0;
 
   gst_video_info_init (info);
   gst_video_alignment_reset (&align);
@@ -4607,18 +4635,57 @@ gst_v4l2_object_acquire_format (GstV4l2Object * v4l2object, GstVideoInfo * info)
     if (v4l2object->ioctl (v4l2object->video_fd, VIDIOC_G_CROP, &crop) >= 0)
       r = &crop.c;
   }
+  GST_INFO_OBJECT (v4l2object->dbg_obj, "crop info: %d,%d,%u,%u",
+      r->left, r->top, r->width, r->height);
+
+  if (v4l2object->downscale) {
+    struct v4l2_selection s_sel;
+    guint wanted_width, wanted_height;
+
+    // set selection with downscale size
+    gst_v4l2_object_adjust_downscale_size (v4l2object, &fmt, r, &wanted_width,
+        &wanted_height);
+    GST_INFO_OBJECT (v4l2object->dbg_obj, "(top, left) adjusted to %dx%d",
+        r->left, r->top);
+
+    memset (&s_sel, 0, sizeof (struct v4l2_selection));
+    s_sel.type = v4l2object->type;
+    s_sel.target = V4L2_SEL_TGT_COMPOSE;
+    s_sel.flags = V4L2_SEL_FLAG_GE;
+    s_sel.r.left = 0;
+    s_sel.r.top = 0;
+    s_sel.r.width = wanted_width;
+    s_sel.r.height = wanted_height;
+
+    if (v4l2object->ioctl (v4l2object->video_fd, VIDIOC_S_SELECTION,
+            &s_sel) < 0) {
+      GST_WARNING_OBJECT (v4l2object->dbg_obj,
+          "Failed to set downscale size with VIDIOC_S_SELECTION: %s",
+          g_strerror (errno));
+    } else {
+      r->width = s_sel.r.width;
+      r->height = s_sel.r.height;
+      crop_width = s_sel.r.width - v4l2object->downstream_width;
+      crop_height = s_sel.r.height - v4l2object->downstream_height;
+
+      GST_INFO_OBJECT (v4l2object->dbg_obj,
+          "s_selection return %dx%d with extra crop width %d, height %d",
+          s_sel.r.width, s_sel.r.height, crop_width, crop_height);
+    }
+  }
+
   if (r) {
     align.padding_left = r->left;
     align.padding_top = r->top;
     align.padding_right =
-        (gint) width - (gint) r->width - r->left - v4l2object->crop_width;
+        (gint) width - (gint) r->width - r->left + crop_width;
     align.padding_bottom =
-        (gint) height - (gint) r->height - r->top - v4l2object->crop_height;
-    width = (gint) r->width + v4l2object->crop_width;
-    height = (gint) r->height + v4l2object->crop_height;
+        (gint) height - (gint) r->height - r->top + crop_height;
+    width = (gint) r->width - crop_width;
+    height = (gint) r->height - crop_height;
     GST_INFO_OBJECT (v4l2object->dbg_obj,
-        "caps size: %ux%u, padding size: %ux%u", width, height,
-        align.padding_right, align.padding_bottom);
+        "caps size: %ux%u, padding size: %u,%u,%u,%u", width, height,
+        align.padding_left, align.padding_right, align.padding_top, align.padding_bottom);
   }
 
   switch (fmt.fmt.pix.field) {
diff --git a/sys/v4l2/gstv4l2object.h b/sys/v4l2/gstv4l2object.h
index c17a012..8d0c28a 100644
--- a/sys/v4l2/gstv4l2object.h
+++ b/sys/v4l2/gstv4l2object.h
@@ -237,8 +237,9 @@ struct _GstV4l2Object {
   gboolean skip_try_fmt_probes;
   gboolean is_amphion;
   gboolean is_hantro;
-  gint crop_width;
-  gint crop_height;
+  guint downstream_width;
+  guint downstream_height;
+  gboolean downscale;
   gboolean seek;
 
   guint max_width;
diff --git a/sys/v4l2/gstv4l2videodec.c b/sys/v4l2/gstv4l2videodec.c
index 00e1614..b96ae87 100644
--- a/sys/v4l2/gstv4l2videodec.c
+++ b/sys/v4l2/gstv4l2videodec.c
@@ -460,6 +460,53 @@ get_hdr10_meta (GstV4l2Object * v4l2object, struct v4l2_hdr10_meta *Hdr10Meta)
   return TRUE;
 }
 
+static void
+gst_v4l2_video_dec_decide_downscale (GstVideoDecoder * decoder)
+{
+  GstV4l2VideoDec *self = GST_V4L2_VIDEO_DEC (decoder);
+  GstV4l2Object *v4l2object = self->v4l2capture;
+  GstCaps *filter, *caps;
+  GstStructure *structure;
+  const GValue *value;
+  gint width = 0, height = 0;
+  gint src_width, src_height;
+
+  src_width = self->v4l2output->info.width;
+  src_height = self->v4l2output->info.height;
+
+  // get the downscale width/height from caps if have
+  filter = gst_caps_new_empty_simple ("video/x-raw");
+  caps = gst_pad_peer_query_caps (decoder->srcpad, filter);
+  structure = gst_caps_get_structure (caps, 0);
+  GST_INFO_OBJECT (v4l2object->dbg_obj, "queried caps: %" GST_PTR_FORMAT, caps);
+
+  value = gst_structure_get_value (structure, "width");
+  if (value && G_VALUE_TYPE (value) == G_TYPE_INT)
+    width = g_value_get_int (value);
+
+  value = gst_structure_get_value (structure, "height");
+  if (value && G_VALUE_TYPE (value) == G_TYPE_INT)
+    height = g_value_get_int (value);
+
+  gst_caps_unref (filter);
+  gst_caps_unref (caps);
+
+  /* not downscale if queried size is a range */
+  if (!width || !height)
+    v4l2object->downscale = FALSE;
+
+  /* still downscale if size info is missing in caps as actual
+   * size has been parsed after source change */
+  if ((!src_width && !src_height) || (width >= src_width / 8
+          && width <= src_width && height >= src_height / 8
+          && height <= src_height)) {
+    v4l2object->downstream_width = width;
+    v4l2object->downstream_height = height;
+    v4l2object->downscale = TRUE;
+    GST_INFO_OBJECT (v4l2object->dbg_obj, "downscale to %dx%d", width, height);
+  }
+}
+
 static gboolean
 gst_v4l2_video_dec_negotiate (GstVideoDecoder * decoder)
 {
@@ -488,6 +535,10 @@ gst_v4l2_video_dec_negotiate (GstVideoDecoder * decoder)
   self->v4l2capture->info.fps_n = self->v4l2output->info.fps_n;
 
   self->v4l2capture->is_g2 = self->v4l2output->is_g2;
+  self->v4l2capture->info.width = self->v4l2output->info.width;
+  self->v4l2capture->info.height = self->v4l2output->info.height;
+
+  gst_v4l2_video_dec_decide_downscale (decoder);
 
   /* For decoders G_FMT returns coded size, G_SELECTION returns visible size
    * in the compose rectangle. gst_v4l2_object_acquire_format() checks both
@@ -831,75 +882,6 @@ gst_v4l2_video_dec_wait_for_src_ch (GstV4l2VideoDec * self)
   return flowret;
 }
 
-static void
-gst_v4l2_video_dec_set_selection (GstVideoDecoder * decoder)
-{
-  GstV4l2VideoDec *self = GST_V4L2_VIDEO_DEC (decoder);
-  GstV4l2Object *v4l2object = self->v4l2capture;
-  struct v4l2_selection sel = { 0 };
-  GstCaps *filter, *caps;
-  GstStructure *structure;
-  const GValue *value;
-  gint width = 0, height = 0;
-  gint src_width, src_height;
-
-  src_width = self->v4l2output->info.width;
-  src_height = self->v4l2output->info.height;
-
-  // get the downscale width/height from caps if have
-  filter = gst_caps_new_empty_simple ("video/x-raw");
-  caps = gst_pad_peer_query_caps (decoder->srcpad, filter);
-  structure = gst_caps_get_structure (caps, 0);
-  GST_INFO_OBJECT (v4l2object->dbg_obj, "queried caps: %" GST_PTR_FORMAT, caps);
-
-  value = gst_structure_get_value (structure, "width");
-  if (value && G_VALUE_TYPE (value) == G_TYPE_INT)
-    width = g_value_get_int (value);
-
-  value = gst_structure_get_value (structure, "height");
-  if (value && G_VALUE_TYPE (value) == G_TYPE_INT)
-    height = g_value_get_int (value);
-
-  /* not downscale if queried size is a range */
-  if (!width || !height)
-    goto out;
-
-  /* still downscale if size info is missing in caps as actual
-   * size has been parsed after source change */
-  if ((!src_width && !src_height) || (width >= src_width / 8
-          && width <= src_width && height >= src_height / 8
-          && height <= src_height)) {
-    GST_INFO_OBJECT (v4l2object->dbg_obj, "want to s_selection %dx%d", width,
-        height);
-
-    // set selection with downscale size
-    sel.type = v4l2object->type;
-    sel.target = V4L2_SEL_TGT_COMPOSE;
-    sel.flags = V4L2_SEL_FLAG_GE;
-    sel.r.left = 0;
-    sel.r.top = 0;
-    sel.r.width = width;
-    sel.r.height = height;
-
-    if (v4l2object->ioctl (v4l2object->video_fd, VIDIOC_S_SELECTION, &sel) < 0) {
-      GST_WARNING_OBJECT (v4l2object->dbg_obj,
-          "Failed to set downscale size with VIDIOC_S_SELECTION: %s",
-          g_strerror (errno));
-      goto out;
-    }
-    v4l2object->crop_width = width - (gint) sel.r.width;
-    v4l2object->crop_height = height - (gint) sel.r.height;
-
-    GST_INFO_OBJECT (v4l2object->dbg_obj, "got downscale size %dx%d",
-        sel.r.width, sel.r.height);
-  }
-
-out:
-  gst_caps_unref (filter);
-  gst_caps_unref (caps);
-  return;
-}
-
 static void
 gst_v4l2_video_dec_loop (GstVideoDecoder * decoder)
 {
@@ -920,8 +902,6 @@ gst_v4l2_video_dec_loop (GstVideoDecoder * decoder)
       goto beach;
     }
 
-    gst_v4l2_video_dec_set_selection (decoder);
-
     GST_DEBUG_OBJECT (decoder, "Setup the capture queue");
     if (G_UNLIKELY (!GST_V4L2_IS_ACTIVE (self->v4l2capture))) {
       if (!gst_video_decoder_negotiate (decoder)) {
-- 
2.43.0

